ddpui/websockets/airbyte_consumer.py

```diff
@@ -123,3 +129,98 @@ def websocket_receive(self, message):
                 status=WebsocketResponseStatus.SUCCESS,
             )
         )
+
+
+class SchemaCatalogConsumer(BaseConsumer):
+    """websocket for checking source schema_catalog"""
+
+    def websocket_receive(consumer, message):
+        logger.info("Recieved the message from client, get schema_catalog inside the connection")
+        payload = json.loads(message["text"])
+
+        orguser: OrgUser = consumer.orguser
+        if orguser.org.airbyte_workspace_id is None:
+            consumer.respond(
+                WebsocketResponse(
+                    data={},
+                    message="Create an airbyte workspace first.",
+                    status=WebsocketResponseStatus.ERROR,
+                )
+            )
+            return
+
+        if "sourceId" not in payload or payload["sourceId"] is None:
+            consumer.respond(
+                WebsocketResponse(
+                    data={},
+                    message="SourceId is required in the payload",
+                    status=WebsocketResponseStatus.ERROR,
+                )
+            )
+            return
+
+        source_id = payload["sourceId"]
+
+        # creating a key and checking if it exists or not
+        task_key = f"{TaskProgressHashPrefix.SOURCE_SCHEMA_CATALOG}-{orguser.org.slug}-{source_id}"
+        task_progress = SingleTaskProgress.fetch(task_key)
+
+        if task_progress is not None:
+            polling_celery(consumer, task_key)
+            return
+
+        # This gives the task to celery
+        get_schema_catalog_task.delay(task_key, str(orguser.org.airbyte_workspace_id), source_id)
```

Comment:
when does this ever happen i.e. when would we find ourselves in this function without an STP

Reply:
So the flow is , 
1. Initially there will be NO STP, so the code will NOT go inside the 168-170 block. 
2. It will get to line 173 and create a STP in celery and then start the polling.

If I close the browser or refresh etc, and come back again to create connection and select the same source again.
1. The previous task would still be running in celery (if not completed)
2. The task_key would be same for the  same source.
4. So we will get a running STP, by the SingleTaskFunction.fetch(task_key) function line 166
3. Now the code block will enter 168-170 and start polling. (It will not give a new task to celery)